pypeit/spectrographs/keck_nirspec.py

```diff
@@ -336,12 +1422,53 @@ def bpm(self, filename, det, shape=None, msbias=None):
 
         return bpm_img
 
-class KeckNIRSPECLowSpectrograph(KeckNIRSPECSpectrograph):
+
+class KeckNIRSPECLowSpectrographOld(KeckNIRSPECLowSpectrograph):
```

Comment:
I think this is incorrect. `KeckNIRSPECLowSpectrograph` is the post-upgrade class.

Overall, I think this code needs to be refactored to have two parent classes (one for pre-upgrade, and one for post-upgrade) and this should contain information about the detector and any functions that are independent of the echelle/longslit modes. Then, each of the parent classes will have two children (one for echelle and one for longslit). You don't necessarily need the first parent class that is listed at the top of this file. Instead, just have the two parents and 4 children. This ensures that we don't have too much duplicated code to maintain.

Reply:
@rcooke-ast Just so you know, the structure of this code here is the result of my (maybe poor) attempt at refactoring. I added `KeckNIRSPECLowSpectrographOld` for completeness, but it is not supported.

Reply:
I'm not sure how many differences there are between the high and low spectrograph settings but I know that the detector parameters will be different at least because of the different orientation of the trace. So there will probably need to be a lot of repeated code for the 4 options anyway. I'll check with Greg about that.